### DISCUSSION TITLE
dpi on vipsthumbnail cli output

### DISCUSSION THREAD

**Discussion by jrochkind** [libvips/libvips#3567] [Q&A] [answered]
I have a large TIFF whose metadata says it's DPI is 400. 

If I create a jpeg with `vipsthumbnail input.tiff --size 1200x65500 -o out-smaller.tiff`:

Then the DPI metadata on the output  jpeg is preserved at 400.  But this isn't quite right now, of course, since i've shrunk the image, it actually has a smaller DPI.  This matters if using a viewer that tries to display the image at "correct size" using DPI, or putting into another non-vips pipeline that is going to use dpi to correctly size a PDF, etc. 

If I create a jpeg200 with `vipsthumbnail input.tiff --size 1200x65500 -o out-smaller.tiff`, it's even more basic:  the dpi seems set to 72, always, no matter the dpi of the input. 

I if I know what dpi I want it to have in metadata, is there any way I can manually specify the dpi metadata I want on the output of `vipsthumbnail`?  Or another vips command or set of vips command I can use to resize an image (and convert colors with `--export-profile rgb`)?

Thanks for vips, and for any ideas.

**Comment by jcupitt** [comment, marked as answer]
Hi @jrochkind,

DPI is awkward because its meaning is not well defined. libvips carries it around, but never changes it -- that's supposed to be a job for the application layer, which probably knows how DPI should be interpreted in this specific case.

As a mostly standalone tool, `vipsthumbnail` is the application layer, so I agree that it probably should set resolution for you. Most people would expect it to, anyway.

As a workaround, you'd probably need to use python etc. rather than `vipsthumbnail`.

**Reply by jrochkind** [reply]
ok thank you for response and info!  In the current use-case I'm in, I think I have found a way to work around it with other tools I already had involved. 

If I were using vips via ruby or python bindings, it would give me tools to set the dpi myself on the output? 

(In the jp2 case, it did not seem to carry it around, but reset it to 72dpi -- so perhaps that's a bug, or OpenJPEG deficiency? I wonder if that means even if vips as a library would ordinarily give me those tools, it might not work on jp2)

I can see how determining the "right" dpi might get tricky; if `vipsthumbnail` just gave a `--dpi` argument or similar that lets the operator, who knows how dpi should be interpreted, set it, that would be a step up!

**Reply by jcupitt** [reply]
Yes, in python (for example) it's:

```python
#!/usr/bin/env python

import pyvips

filename = "/home/john/pics/k2.jpg"
thumb = pyvips.Image.thumbnail(filename, 600)

image = pyvips.Image.new_from_file(filename)
old_width = image.width
new_width = thumb.width
resize_factor = new_width / old_width
thumb = thumb.copy(xres=thumb.xres * resize_factor, 
                   yres=thumb.yres * resize_factor)

thumb.write_to_file("x.jpg")
```

**Reply by jcupitt** [reply]
openjpeg doesn't support setting resolution in jp2 files, or perhaps I just didn't get around to implementing it, I can't remember. It's missing ICC profile support too.

**Reply by jrochkind** [reply]
OK thanks, that's helpful!

Since in this case my target is jp2, even using vips as a library wouldn't help me!  I perhaps could use something like exiftool to try to set the metadata (haven't tried to confirm that tool specifically can set this, on an jp2) 

-- but in fact, my ultimate target is embedding the jp2 in a PDF, for which I'm currently using the python `img2pdf`, which lets me set the PDF page size independently of the jp2 dpi, in a way that seems like the jp2 dpi becomes irrelevant. (If I don't override the dpi, img2pdf uses the (incorrect) jp2 dpi to wrongly set the PDF page size, which is where I first noticed the issue). 

I am realizing that jp2 is just kind of an under-served format, even with OpenJPEG (in general, not in vips specifically).  

If at some point you had time/interest to see if vips could have it's jp2 writing enhanced to be at the level of other writers (or even file issues on OpenJPEG asking for the features necessary to do so) -- it would be very helpful!  But I realize you have your own priorities (as do other users!) and time management, and I really appreciate everything you do for vips, it's an amazing tool, and your user support is so appreciated! 

(Some parts of some OpenJPEG docs I found suggest to me it has some kind of support for ICC profiles, but perhaps not complete).

**Reply by jcupitt** [reply]
Yes, jp2k is not much loved or used :( It's slightly better than plain JPEG, but adds some nasty patent issues, and various rights holders who tried to extract money from users. There's still no high quality, free reference implementation.

I don't use it myself, so I don't have much incentive to improve support. PRs welcome, of course!

**Reply by jrochkind** [reply]
I think the patent issues and any other IP issues are resolved, but it has never been able to get over it's bad reputation in that area, as you have repeated. 

It turns out it's pretty universally supported by PDF readers, and you can make much smaller PDFs (for same perceived image quality) using jp2 than jpeg or png. Is how I come to messing with it. 

But it is hard to work with with so little software support.  Which at this point is unlikely to change. Sad but true. 

I'm glad that vips can do what it can do, at least! 

Writing image procesing code in C is well outside my area of expertise, but perhaps at some point some academic and non-profit institutions where it still remains of interest could get it together to pay someone (or you) to PR.
